File: backend/app/models/song.py
from pydantic import BaseModel
from typing import Optional


class SongMetadata(BaseModel):
    key: str
    scale: str
    bpm: int
    timeSignature: str
    recommendedCapo: int


class MediaInfo(BaseModel):
    sourceType: str
    sourceUrl: str
    audioDuration: float
    audioUrl: str = ""


class ChordEvent(BaseModel):
    timestamp: float
    chord: str
    lyrics: str


class Song(BaseModel):
    id: str
    title: str
    artist: str
    metadata: SongMetadata
    media: MediaInfo
    chordTimeline: list[ChordEvent]
    tablatureData: str

File: frontend/src/types/song.ts
export interface SongMetadata {
  key: string;
  scale: "major" | "minor";
  bpm: number;
  timeSignature: string;
  recommendedCapo: number;
}

export interface MediaInfo {
  sourceType: "youtube" | "upload" | "search";
  sourceUrl: string;
  audioDuration: number;
  audioUrl?: string;
}

export interface ChordEvent {
  timestamp: number;
  chord: string;
  lyrics: string;
}

export interface Song {
  id: string;
  title: string;
  artist: string;
  metadata: SongMetadata;
  media: MediaInfo;
  chordTimeline: ChordEvent[];
  tablatureData: string;
}

export interface TrendingTab {
  id: string;
  title: string;
  artist: string;
  key: string;
  tempo: number;
  difficulty: "beginner" | "intermediate" | "advanced";
}
